hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -106,33 +157,60 @@ impl FlatGraphBuilder {
                 }
             }
             HfStatement::Pipeline(pipeline_stmt) => {
-                self.add_pipeline(pipeline_stmt.pipeline, None);
+                if let Err(err) = self.add_pipeline(pipeline_stmt.pipeline, None) {
+                    self.diagnostics.push(err.into());
+                }
             }
         }
     }
 
     /// Helper: Add a pipeline, i.e. `a -> b -> c`. Return the input and output ends for it.
-    fn add_pipeline(&mut self, pipeline: Pipeline, current_varname: Option<&Ident>) -> Ends {
+    fn add_pipeline(
+        &mut self,
+        pipeline: Pipeline,
+        current_varname: Option<&Ident>,
+    ) -> syn::Result<Ends> {
```

Comment:
Don't need to return the `syn::Result`, just do `Result<Ends, ()>` and push to `self.diagnostics` instead

Reply:
Or can even keep returning just `Ends` and keep going even when errors happen, which is how things work for bad ports right now. depends on if the cascading error messages are meaningful or not